wikimedia/wikipedia-ios
# Update Push notifications settings view

**Phabricator:** [T288688 ](https://phabricator.wikimedia.org/T288688) and partially [T288773](https://phabricator.wikimedia.org/T288773)

### Notes
- Updates the Push notifications settings view to support enabling system alert/badge permissions and subscribing/unsubscribing to echo notifications. 
- Shows/hides the "Push notifications" row in the root Settings view depending on if the user is authenticated or not
- Unsubscribes device from echo notifications when user logs out of account
- Also adds the Settings related edge and error case work from [T288773](https://phabricator.wikimedia.org/T288773).

### Test Steps

Updating `PushNotificationsSettingsSwitchItem` titles in `private func updateSections()` in `PushNotificationsSettingsViewController.swift` might make it easier to visualize the current state of the view. For example, something like:
```swift
let pushSwitchItem = PushNotificationsSettingsSwitchItem(title: "\(status) - \(UserDefaults.standard.wmf_isSubscribedToEchoNotifications)" ...
```

1. Would be best to delete and fresh install (or reset permissions) on a test device to see the system push alert permission modals. You will not be able to properly test this on the iOS Simulator because it doesn't provide a device token.
2. Sign out of your account. Confirm "Push notifications" isn't listed in the Settings root view.
3. Sign in to an account. Confirm "Push notifications" now appears.
4. Enable and disable the switch, confirm it lines up with expectations (general criteria is listed in the development notes in the description of https://phabricator.wikimedia.org/T288688)
5. With the "Push notifications" toggle on, log out of your account.
6. Sign back in, and confirm when returning to the "Push notifications" view the switch is off.

In order to test receiving actual notifications, we'll have to test this out in a TestFlight build.
